### PR TITLE
Add alternative receiver for transfer in erc20 to native mode

### LINKED ISSUE #296: Alternative receiver in ERC20-to-NATIVE mode
This issue is to implement suggestions from #295 for the contracts related to the `erc20-to-native` bridge mode

### PR DESCRIPTION
**Pull request by patitonar**
Closes #296

### CODE REVIEW

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol [hunk 1]
```diff
@@ -5,38 +5,52 @@ import "../ERC20Bridge.sol";
 
 contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge {
     event RelayedMessage(address recipient, uint256 value, bytes32 transactionHash);
+    event UserRequestForAffirmation(address recipient, uint256 value);
+
+    bytes32 internal constant BRIDGE_CONTRACT = keccak256(abi.encodePacked("bridgeOnOtherSide"));
```
[the remark this answers is not in the document]
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Fixed here https://github.com/poanetwork/poa-bridge-contracts/pull/294/commits/e0b3ad6c9a04676b88616beb79c23ab3f74ab75f

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol [hunk 2]
```diff
@@ -5,38 +5,52 @@ import "../ERC20Bridge.sol";
...
+        uintStorage[EXECUTION_MAX_PER_TX] = _homeDailyLimitHomeMaxPerTxArray[1];
         uintStorage[DECIMAL_SHIFT] = _decimalShift;
         setOwner(_owner);
+        _setBridgeContractOnOtherSide(_bridgeOnOtherSide);
         setInitialize();
 
         emit RequiredBlockConfirmationChanged(_requiredBlockConfirmations);
         emit GasPriceChanged(_gasPrice);
-        emit ExecutionDailyLimitChanged(_maxPerTxHomeDailyLimitHomeMaxPerTxArray[1]);
+        emit ExecutionDailyLimitChanged(_homeDailyLimitHomeMaxPerTxArray[0]);
```
[the remark this answers is not in the document]
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Thanks, fixed https://github.com/poanetwork/poa-bridge-contracts/pull/294/commits/1280235df900f4fe3ca1c7f35618688515ca2dfc

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol [hunk 3]
```diff
@@ -5,38 +5,52 @@ import "../ERC20Bridge.sol";
...
-        uint256 _decimalShift
+        uint256 _decimalShift,
+        address _bridgeOnOtherSide
     ) external returns (bool) {
         require(!isInitialized());
         require(AddressUtils.isContract(_validatorContract));
         require(_requiredBlockConfirmations != 0);
         require(_gasPrice > 0);
-        require(_maxPerTxHomeDailyLimitHomeMaxPerTxArray[2] < _maxPerTxHomeDailyLimitHomeMaxPerTxArray[1]); // _homeMaxPerTx < _homeDailyLimit
+        require(
```
[the remark this answers is not in the document]
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: When deploying the contracts, having 3 `require`s increases the gas usage but when making a call to the method the gas usage difference is very small in favor of having 3 `require` which consume less gas. Given the size of our contracts, I think we should keep having only one `require`

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol [hunk 4]
```diff
@@ -63,4 +77,33 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge {
...
+        require(_amount > 0);
+        require(withinLimit(_amount));
+        setTotalSpentPerDay(getCurrentDay(), totalSpentPerDay(getCurrentDay()).add(_amount));
+
+        erc20token().transferFrom(_sender, address(this), _amount);
+        emit UserRequestForAffirmation(_receiver, _amount);
+    }
+
+    function relayRequest(address _from, address _receiver, uint256 _amount) external {
+        require(_from == msg.sender || _from == _receiver);
```
[the remark this answers is not in the document]
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: The idea of this check was to allow only the user who approved the token transfer to specify a different receiver for the other network. If it is not present, after the user calling `approve` method, someone could call `relayRequest` using his address as receiver, so tokens will be transferred to the bridge contract, but native tokens will be minted to the attacker's account

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: OK. Thanks

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Got you! Thanks!

So, please advise where (contracts or oracle) we need to add the following tests:
- A makes `approve` and invokes `relayRequest` with A as the recipient
- A makes `approve` and B invokes `relayRequest` with A as the sender and A as the recipient 
- A makes `approve` and B invokes `relayRequest` with A as the sender and B as the recipient 
- A makes `approve` and B invokes `transferFrom` with A as the sender and the bridge contract as the recipient

[the remark this answers is not in the document]
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Updated the unit tests to cover all the scenarios mentioned. https://github.com/poanetwork/poa-bridge-contracts/pull/294/commits/e9d6487f5a2fd6f570af826fd30293ccab71a7d2 

The last scenario is not possible, `transferFrom` checks that the `msg.sender` is the address that was approved, so if A call `approve` to the bridge contract then only the bridge contract can call `transferFrom` directly to move those tokens

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/0e65947efbffc592cffea8c2ae9d3b8e11659854/contracts/token/ERC20/StandardToken.sol#L34

[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Thanks, indeed it is covered in the ERC20 standard: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md#transferfrom.

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol [hunk 5]
```diff
@@ -63,4 +78,33 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge {
...
+        require(_receiver != bridgeContractOnOtherSide());
+        require(_amount > 0);
+        require(withinLimit(_amount));
+        setTotalSpentPerDay(getCurrentDay(), totalSpentPerDay(getCurrentDay()).add(_amount));
+
+        erc20token().transferFrom(_sender, address(this), _amount);
+        emit UserRequestForAffirmation(_receiver, _amount);
+    }
+
+    function relayRequest(address _from, address _receiver, uint256 _amount) external {
```
Comment (akolotov) [inline contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Do you think it makes sense to be consistent with the contracts implementing `erc677-to-erc677` on top of AMB and name the method `relayTokens`?

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: > Updated the unit tests to cover all the scenarios mentioned.

@patitonar sorry but I cannot find these tests. The idea was to check behavior of `relayRequest` if the sender differs from the approver. But the tests operates under the same account that approved the transfer.

[the remark this answers is not in the document]
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: By bad, all tests weren't visible from the diff of the commit. Here are the lines where the scenarios are tested:

- A makes approve and invokes relayRequest with A as the recipient
https://github.com/poanetwork/poa-bridge-contracts/blob/e9d6487f5a2fd6f570af826fd30293ccab71a7d2/test/erc_to_native/foreign_bridge.test.js#L684

- A makes approve and B invokes relayRequest with A as the sender and A as the recipient
https://github.com/poanetwork/poa-bridge-contracts/blob/e9d6487f5a2fd6f570af826fd30293ccab71a7d2/test/erc_to_native/foreign_bridge.test.js#L734-L735

- A makes approve and B invokes relayRequest with A as the sender and B as the recipient
https://github.com/poanetwork/poa-bridge-contracts/blob/e9d6487f5a2fd6f570af826fd30293ccab71a7d2/test/erc_to_native/foreign_bridge.test.js#L732

[the remark this answers is not in the document]
Comment (akolotov) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: thanks!

File: contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol (hunk 5 above)
Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Yes, I think we should be consistent. We are also using `relayRequest` for alternative receiver in the native side.  What do you think would be best, keeping `relayRequest` for native side and name `relayTokens` on erc20 sides or having the same name `relayRequest` for both native and erc20 sides? In the second case the name in  `erc677-to-erc677` on top of AMB should be updated

Comment (patitonar) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: Thanks. Renamed `relayRequest` to `relayTokens` https://github.com/poanetwork/poa-bridge-contracts/pull/294/commits/4fb9b9eb6c73fc71db489620879c79cf6865c644

I'll apply similar changes to the other PRs so `relayTokens` is used

Comment (akolotov) [reply contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol]: thanks

**Pull request merged**